---
title: "From miRNAs to genes"
author: "Pilar Gonz√°lez Marchante"
output: 
  rmarkdown::github_document:
    toc: true
    toc_depth: 2
---

```{r setup, include=FALSE}
```

```{r, eval = FALSE}
write.table(rownames(mirna.filt.counts), file = "results/preprocessing/cookingmiRNASeq/mirna.ids.txt", sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)
```


# Associating miRbase IDs to Gene Symbols with miRWalk

We discovered which miRNAs regulate which genes using online tool miRWalk, which allows us to retrieve only those associations that have been experimentally validated, since they appear in miRTarBase (not just predicted *in silico*) and that are on both TargetScan and miRDB, which computationally predict their targets.

![](miRWalk.png)

```{r, eval = FALSE}
mirwalk.1 <- read.csv("scripts/associations/miRNA-gene/miRWalk_miRNA_Targets.part1.csv")
mirwalk.2 <- read.csv("scripts/associations/miRNA-gene/miRWalk_miRNA_Targets.part2.csv")
miRNA2gene <- rbind(mirwalk.1, mirwalk.2)
miRNA2gene <- miRNA2gene[, c(1,3)]
miRNA2gene <- unique(miRNA2gene)
```


```{r, eval = FALSE}
head(miRNA2gene)

#          mirnaid genesymbol
# 1  hsa-let-7a-5p        RDX
# 2  hsa-let-7a-5p    ANKRD46
# 4  hsa-let-7a-5p       YOD1
# 5  hsa-let-7a-5p       MDM4
# 13 hsa-let-7a-5p      SURF4
# 14 hsa-let-7a-5p      AP1S1
```

# Converting Gene Symbols to ENSEMBL IDs

However, there is one problem: our gene expression matrix comes in the form of ENSEMBL IDs, yet this website gives genes as Gene Symbols. We have one more task to complete: to convert Gene Symbols to ENSEMBL IDs, with the help of `biomaRt`.

```{r, eval = FALSE}
gene.symbols <- unique(miRNA2gene$genesymbol) # we need to map 1385 genes
library(biomaRt)
mart = useMart(biomart="ensembl", dataset="hsapiens_gene_ensembl")
gene.conversion <- getBM(filters="external_gene_name",
              attributes=c("ensembl_gene_id", "external_gene_name"), 
              values= gene.symbols,
              mart=mart)
```

We have two problems to deal with: missing genes and genes with multiple ENSEMBL IDs. For the missing genes part, we can input them into [this multisymbol checker](https://www.genenames.org/tools/multi-symbol-checker/) to get the latest version of the gene name. For duplicated genes, what we did was to keep the main sequence (which happens to always be the last element) and delete the alternative ones (haplotypes/patches).

```{r, eval = FALSE}
gene.conversion <- gene.conversion[order(gene.conversion$external_gene_name), ]
rownames(gene.conversion) <- c(1:length(gene.conversion$ensembl_gene_id))

duplicated.genes <- which(duplicated(gene.conversion$external_gene_name))
duplicated.genes <- duplicated.genes-1 # we keep the last one, as this is the main sequence as opposed to alternative sequences (haplotypes/patches)
gene.conversion <- gene.conversion[-duplicated.genes, ]

missing.genes <- setdiff(gene.symbols, gene.conversion$external_gene_name) # 2
write.table(missing.genes, file = "scripts/associations/miRNA-gene/missing.genes.txt", sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)

missing.genes <- read.table(file = "scripts/associations/miRNA-gene/hgnc-symbol-check.csv", sep = ",")
colnames(missing.genes) <- missing.genes[1, ]
missing.genes <- missing.genes[-1, ]

missing.gene.conversion <- getBM(filters="external_gene_name",
              attributes=c("ensembl_gene_id", "external_gene_name"),
              values=missing.genes$`Approved symbol`,
              mart=mart)

missing.gene.conversion$external_gene_name <- missing.genes$Input[match(missing.gene.conversion$external_gene_name, missing.genes$`Approved symbol`)] # replace new names by old names, so that it finds them in the associations

gene.conversion <- rbind(gene.conversion, missing.gene.conversion)
```

As expected, some gene symbols map to several ENSEMBL IDs. We decided to map everything into ENSEMBL instead of going from ENSEMBL to Entrez/Gene Symbol because some ENSEMBL transcripts are novel or putative so they haven't been given an external ID yet. If we converted from ENSEMBL to Entrez, we would have to remove genes from the expression matrix; but if we convert from Entrez/Symbol to ENSEMBL, we just have to add extra associations to the associations matrix, which is fine.

```{r, eval = FALSE}
miRNA2gene$ensembl_gene_id <- gene.conversion$ensembl_gene_id[match(miRNA2gene$genesymbol, gene.conversion$external_gene_name)]
```

# Associating mature miRNAs to their precursor genes

In order to visualize miRNA expression values in Paintomics pathways, we need to add "fake associations" between the mature miRNA and the precursor miRNA gene from which it originated.

```{r, eval = FALSE}
load(file = "data/cooked/miRNA-Seq/miRNA.filt.rda")
library(miRBaseConverter)
precursors <- miRNA_MatureToPrecursor(miRNANames = rownames(mirna.filt.counts))
accessions <- miRNA_NameToAccession(miRNANames = precursors$Precursor, version = "v20")
precursors$accession <- accessions$Accession[match(precursors$Precursor, accessions$miRNAName_v20)]

library(biomaRt)
mart = useMart(biomart="ensembl", dataset="hsapiens_gene_ensembl")
length(unique(precursors$accession)) # 448 accessions to convert
genes.miRNAs <- getBM(filters="mirbase_accession",
                         attributes=c("mirbase_accession", "ensembl_gene_id"), 
                         values=precursors$accession,
                         mart=mart)

missing.accessions <- setdiff(accessions$Accession, genes.miRNAs$mirbase_accession)
length(missing.accessions) # 4 missing accessions

genes.miRNAs <- genes.miRNAs[order(genes.miRNAs$mirbase_accession), ]
rownames(genes.miRNAs) <- c(1:length(genes.miRNAs$ensembl_gene_id))

duplicated.genes <- which(duplicated(genes.miRNAs$mirbase_accession))
duplicated.genes <- duplicated.genes-1 # we keep the last one, as this is the main sequence as opposed to alternative sequences (haplotypes/patches)
genes.miRNAs <- genes.miRNAs[-duplicated.genes, ]

missing.accessions
# [1] "MI0000542" "MI0003776" "MI0003778" "MI0008190"

missing.accessions <- data.frame(c("MI0000542", "MI0003776", "MI0003778", "MI0008190"), c("ENSG00000208037", "ENSG00000211543", "ENSG00000221493", "ENSG00000211491"))
colnames(missing.accessions) <- colnames(genes.miRNAs)
genes.miRNAs <- rbind(genes.miRNAs, missing.accessions)

precursors$ensembl_gene_id <- genes.miRNAs$ensembl_gene_id[match(precursors$accession, genes.miRNAs$mirbase_accession)]

mirna2precursorgene <- data.frame(precursors$OriginalName, 0, precursors$ensembl_gene_id)
names(mirna2precursorgene) <- names(miRNA2gene)

miRNA2gene <- rbind(miRNA2gene, mirna2precursorgene)

write.table(miRNA2gene[, -2], file = "results/associations/miRNA-gene/all_miRWalk_miRNA_Targets.tab", sep = "\t", row.names = FALSE, quote = FALSE)
```

As an approximation we'll choose relevant associations as those where a relevant miRNA appears.

```{r, eval = FALSE}
miRNA.associations <- read.table(file = "results/associations/miRNA-gene/all_miRWalk_miRNA_Targets.tab", header = TRUE)
miRNA.DEGs <- read.table(file = "results/preprocessing/cookingmiRNASeq/common.miRNA.DEGs.txt")
relevant.miRNA.associations <- subset(miRNA.associations, mirnaid %in% miRNA.DEGs$V1) # 970 miRNA-gene associations are relevant per this criterion (of which 793 are miRNA-target and the rest are miRNA-precursor)

write.table(relevant.miRNA.associations, file = "results/associations/miRNA-gene/DEG_miRWalk_miRNA_Targets.tab", sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)
```

