---
title: "From proteins to genes"
author: "Pilar Gonz√°lez Marchante"
output: 
  rmarkdown::github_document:
    toc: true
    toc_depth: 2
---

```{r setup, include=FALSE}
```

# Associating antibodies to Gene Symbols

We had already associated the antibodies used to capture peptide signals with their corresponding Entrez IDs, using [MD Anderson's current expanded antibody list](https://www.mdanderson.org/research/research-resources/core-facilities/functional-proteomics-rppa-core/antibody-information-and-protocols.html). Some peptides had more than 1 Entrez IDs associated with them (for example, Aurora-ABC_p_T288_T232_T198 has 6790, 9212 and 6795, corresponding to aurora kinase A, B and C, respectively). We could keep one at random or delete them altogether; as to not overcomplicate things, we decided to remove all instances with more than 2 Entrez IDs; and for those with 2 Entrez IDs, we kept the first one.

Since we did this before performing differential expression analysis, the code can be seen in `cookingProt.rmd`. 

# Converting Entrez to ENSEMBL IDs

```{r, eval = FALSE}
dea.limma.prot <- read.table(file = "results/preprocessing/cookingProt/limma.ordered.tsv")
library(biomaRt)
mart = useMart(biomart="ensembl", dataset="hsapiens_gene_ensembl")
gene.conversion <- getBM(filters="entrezgene_id",
              attributes=c("ensembl_gene_id", "entrezgene_id", "entrezgene_accession"), 
              values= rownames(dea.limma.prot),
              mart=mart)

which(duplicated(gene.conversion$entrezgene_id))
```


```{r, eval = FALSE}
#  [1]   3  33  34  35  36  37  38  40  67 128 129 130 131 132 133 134 135 136 137 138 139 140 144 145 146 147 173 179 180 186 198
# [32] 226 227 228 229 230 231 245 251 263 271 311 318 331 364 381 416
```


```{r, eval = FALSE}
gene.conversion <- gene.conversion[order(gene.conversion$entrezgene_id), ]

rownames(gene.conversion) <- gene.conversion$ensembl_gene_id
```

There are 369 Entrez IDs in our DEA results, but after converting them to ENSEMBL we got 416. How come? For example, for Entrez gene 31 named ACACA, there are 2 ENSEMBL IDs: ENSG00000275176 (Location: CHR_HSCHR17_7_CTG4:37,086,456-37,411,442) and ENSG00000278540 (Location: 17:37,084,992-37,406,836). The second one comes from the main chromosome (number 17) whereas the other one comes from haplotypic regions, also called alternate locus, which were added as a way to portray individual variability, making it less relevant in our analysis. In fact, ENSG00000275176 wasn't even in our original counts matrix (of 60,660 genes), unlike ENSG00000278540, which was in the original matrix and even remained after filtering. Thus, in cases like this it makes sense to use the gene coming from the main chromosome, instead of the haplotypic one.

```{r, eval = FALSE}
gene.conversion <- gene.conversion[-c(2, 32, 33, 34, 35, 36, 37, 38, 39, 66, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 143, 144, 145, 146, 172, 178, 179, 185, 197, 225, 226, 227, 228, 229, 230, 244, 250, 262, 270, 310, 317, 330, 363, 380, 415), ]
```

We are missing one gene.

```{r, eval = FALSE}
setdiff(rownames(dea.limma.prot), gene.conversion$entrezgene_id) # 780
```

For some reason DDR1 didn't make it to the list, even though it has an ENSEMBL ID of ENSG00000204580.

```{r, eval = FALSE}
missing.gene <- data.frame("ENSG00000204580", "780", "DDR1")
names(missing.gene) <- c("ensembl_gene_id", "entrezgene_id", "entrezgene_accession")
rownames(missing.gene) <- "ENSG00000204580"
gene.conversion <- rbind(gene.conversion, missing.gene)

write.table(gene.conversion, file = "results/associations/protein-gene/entrez2ensembl.txt", sep = "\t", quote = FALSE, row.names = FALSE)
```

# Replacing gene names in DEA results

And we can finally change the names of the genes tested for differential expression analysis into ENSEMBL IDs.

```{r, eval = FALSE}
dea.limma.prot <- read.table(file = "results/preprocessing/cookingProt/limma.ordered.tsv")
dea.limma.prot <- dea.limma.prot[order(as.numeric(row.names(dea.limma.prot))), ]
gene.conversion <- gene.conversion[order(as.numeric(gene.conversion$entrezgene_id)), ]

rownames(dea.limma.prot) == gene.conversion$entrezgene_id

rownames(dea.limma.prot) <- gene.conversion$ensembl_gene_id

prot.expression <- as.data.frame(dea.limma.prot[, 1])
rownames(prot.expression) <- rownames(dea.limma.prot)
names(prot.expression) <- "logFC"

write.table(prot.expression, file = "results/associations/protein-gene/prot.expression.txt", sep = "\t", quote = FALSE, col.names = FALSE)
```

Now we do the same for the significant proteins.

```{r, eval = FALSE}
prot.DEGs <- read.table(file = "results/preprocessing/cookingProt/prot.DEGs.txt")
names(prot.DEGs) <- "entrezgene_id"

prot.DEGs <- merge(prot.DEGs, gene.conversion, by = "entrezgene_id")
prot.DEGs <- prot.DEGs[, 2]

write.table(prot.DEGs, file = "results/associations/protein-gene/prot.DEGs.tab", sep = "\t", quote = FALSE, col.names = FALSE, row.names = FALSE)
```

